File: 205.isomorphic-strings.py
#
# @lc app=leetcode id=205 lang=python3
#
# [205] Isomorphic Strings
#

# @lc code=start
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        hash_table = {}
        seen = set()
        for i in range(len(s)):
            if ((s[i] not in hash_table and t[i] in seen) or (s[i] in hash_table and t[i] != hash_table[s[i]])): return False
            hash_table[s[i]] = t[i]
            seen.add(t[i])            
        return True
# @lc code=end

File: 205.isomorphic-strings.cpp
/*
 * @lc app=leetcode id=205 lang=cpp
 *
 * [205] Isomorphic Strings
 */

// @lc code=start
#include <unordered_map>
#include <unordered_set>
#include <string>

class Solution {
public:
    bool isIsomorphic(std::string s, std::string t) {
        std::unordered_map<char, char> hashTable;
        std::unordered_set<char> seen;
        for (int i = 0; i < s.length(); i++) {
            if ((!hashTable.count(s[i]) && seen.count(t[i])) || (hashTable.count(s[i]) && t[i] != hashTable[s[i]])) {
                return false;
            }
            hashTable[s[i]] = t[i];
            seen.insert(t[i]);
        }
        return true;
    }
};

// @lc code=end

